classes/manager.php

```diff
@@ -163,12 +168,137 @@ public static function get_reasons(float $duration): int {
+
+        // Grab the fastest profile for this page/request, and use that as
+        // the lower boundary for any new profiles of this page/request.
+        $cachekey = $request;
+        $cachefield = "min_duration_for_reason_$reason";
+        $cache = \cache::make('tool_excimer', 'request_metadata');
+        $result = $cache->get($cachekey);
+        if ($result === false || !isset($result[$cachefield])) {
+            // NOTE: Opting to query this way instead of using MIN due to
+            // the fact valid profiles will be added and the limits will be
+            // breached for 'some time'. This will keep the constraints as
+            // correct as possible.
+            $reasons = $DB->sql_bitand('reason', $reason);
+            $sql = "SELECT duration as min_duration
+                      FROM {tool_excimer_profiles}
+                     WHERE $reasons != ?
+                           AND request = ?
+                  ORDER BY duration DESC
+                     ";
+            $resultset = $DB->get_records_sql($sql, [
+                self::REASON_NONE,
+                $request,
+            ], $pagequota - 1, 1); // Will fetch the Nth item based on the quota.
+            // Cache the results in milliseconds (avoids recalculation later).
+            $minduration = (end($resultset)->min_duration ?? 0) * 1000;
+            $result[$cachefield] = $minduration;
+            $cache->set($cachekey, $result);
+        }
+        return $result[$cachefield];
+    }
+
+    /**
+     * Returns the minimum duration for profiles matching this reason.
+     *
+     * Cost: Should be free as long as the cache exists in the config.
+     * Otherwise: 1 DB read, 1 cache write
+     *
+     * @param  int $reason - the profile type or REASON_*
+     * @return float duration (in milliseconds) of the fastest profile for a given reason.
+     */
+    public static function get_min_duration_for_reason(int $reason): float {
+        global $DB;
+
+        $reasonstr = self::REASON_STR_MAP[$reason];
+        $quota = (int) get_config('tool_excimer', "num_$reasonstr");
+
+        $cachekey = 'profile_type_' . $reason . '_min_duration_ms';
+        $result = get_config('tool_excimer', $cachekey);
+        if ($result === false) {
+            // Get and set cache.
+            $reasons = $DB->sql_bitand('reason', $reason);
+            $sql = "SELECT duration as min_duration
+                      FROM {tool_excimer_profiles}
+                     WHERE $reasons != ?
+                  ORDER BY duration DESC
+                     ";
+            $resultset = $DB->get_records_sql($sql, [
+                self::REASON_NONE,
+            ], $quota - 1, 1); // Will fetch the Nth item based on the quota.
+            // Cache the results in milliseconds (avoids recalculation later).
+            $result = (end($resultset)->min_duration ?? 0) * 1000;
+            set_config($cachekey, $result, 'tool_excimer');
```

Comment:
You set the config here if it has not already been set. However, I do not see anywhere where this config gets reset. Once set the first time, it doesn't change. Even after the max number gets reached, I can still add more, and debugger runs shows the config value to remain '0'.

Reply:
Thanks for spotting this, yes I forgot to clear the cache for `set_config` calls in the appropriate places.